### ISSUE TITLE
Cilium assigns additional IP to `cilium_host` upon restart

### ISSUE THREAD

**Issue by anfernee** [cilium/cilium#16520]
## Bug report

**General Information**

- Cilium version (run `cilium version`): Head of master
- Kernel version (run `uname -a`): Current test image: (Linux k8s1 5.12.0-rc4+ #1 SMP Wed Apr 21 11:28:44 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux)
- Orchestration system version in use (e.g. `kubectl version`, ...): 1.21.1

**How to reproduce the issue**
I don't know how exactly to reproduce the issue. However, I found after running e2e test under `test` folder several times, or sometimes the first provision run. e.g. something like the following

```
ginkgo --focus "K8sXXTest" -v -- --cilium.provision=false
```

I will hit some healthcheck timeout failures that is unrelated to the test I am running. The agent log says something like this:

```
level=debug msg="Cannot find socket" error="stat /var/run/cilium/health.sock: no such file or directory" file-path=/var/run/cilium/health.sock subsys=cilium-health-launcher
level=debug msg="stale identity observed" identity=2 ipAddr=10.0.1.235 oldIdentity=1 subsys=hubble
level=debug msg="stale identity observed" identity=2 ipAddr=10.0.1.235 oldIdentity=1 subsys=hubble
level=debug msg="Cannot find socket" error="stat /var/run/cilium/health.sock: no such file or directory" file-path=/var/run/cilium/health.sock subsys=cilium-health-launcher
level=debug msg="Greeting failed" error="Get \"http://10.0.1.60:4240/hello\": dial tcp 10.0.1.60:4240: connect: connection timed out" host="http://10.0.1.60:4240" ipAddr=10.0.1.60 nodeName=k8s1 path="Via L3" subsys=health-server
level=debug msg="Failed to probe: Get \"http://10.0.1.60:4240/hello\": dial tcp 10.0.1.60:4240: connect: connection timed out" ipAddr=10.0.1.60 nodeName=k8s1 port=4240 subsys=health-server
level=debug msg="Greeting host" host="http://[fd02::175]:4240" ipAddr="fd02::175" nodeName=k8s1 path="Via L3" subsys=health-server
level=debug msg="Controller func execution time: 126.546µs" name=metricsmap-bpf-prom-sync subsys=controller uuid=b255135f-b837-44d3-91b8-773e14173344
```

Then I found on `cilium_host`, it has 2 IP addresses:
```
55: cilium_host@cilium_net: <BROADCAST,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether f2:c2:76:3e:4e:fe brd ff:ff:ff:ff:ff:ff
    inet 10.0.0.214/32 scope link cilium_host
       valid_lft forever preferred_lft forever
    inet 10.0.1.100/32 scope link cilium_host
       valid_lft forever preferred_lft forever
    inet6 fd04::11/128 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::f0c2:76ff:fe3e:4efe/64 scope link 
```

This will cause the route to the other node cannot be added. I suspect this is due to the change of PodCIDR to the node, and when it changes, the node fails to cleanup the IP address it added to `cilium_host`. If I look at the node object:

```
- apiVersion: v1
  kind: Node
  metadata:
    annotations:
      io.cilium.network.ipv4-cilium-host: 10.0.0.214
      io.cilium.network.ipv4-health-ip: 10.0.0.80
      io.cilium.network.ipv4-pod-cidr: 10.0.0.0/24
```

ipv4-health-ip changes every time I redeploy cilium. Most of time `ipv4-pod-cidr` and `ipv4-cilium-host` don't change. However, when `ipv4-pod-cidr` changes, it will see the issues I described. 

The issue could cause some test flakiness. I didn't have time to dig into why `ipv4-pod-cidr` will change, though. FWIW, I changed cilium config back and forth between tunnel mode and direct routing mode frequently.

**Comment by aanm** [comment]
I think this was fixed by https://github.com/cilium/cilium/pull/16307 cc @christarazi

**Comment by christarazi** [comment]
@anfernee Did your master build include beb8bde? If so, then we may have another bug related to the pod CIDR changing.

**Comment by anfernee** [comment]
> @anfernee Did your master build include [beb8bde](https://github.com/cilium/cilium/commit/beb8bdea384fdc4ccb10769142c8981bb10334d5)? If so, then we may have another bug related to the pod CIDR changing.

Yes, I do see the change in my tree.

**Comment by christarazi** [comment]
Thanks. I've updated the title of the issue.

**Comment by Purushotham233** [comment]
Seeing the same issue in our test env.
Cilium Version: `1.10.0`
k8s version: `1.20` and `1.19`

A Few more observations:
- Observing connectivity for all pods on the node when the node's cilium-router IP is different from the default route added in the Pod netns.

**Comment by Purushotham233** [comment]
@christarazi fix in #16672 does it removes the additional IPs assigned to the cilium_host interface? Or a separate clean-up is needed?
